### PR TITLE
New ref ax version restrictions

### PR DESCRIPTION
**Pull request by tympollack**

### CODE REVIEW

File: src/main/groovy/com/cedarsoftware/ncube/NCubeManager.groovy [hunk 1]
```diff
@@ -785,22 +793,64 @@ class NCubeManager implements NCubeMutableClient, NCubeTestServer
...
-        // Make sure we are not lock blocked on any of the appId's that are being updated.
-        for (ApplicationID appId : uniqueAppIds)
+    private boolean validateReferenceAxesAppIds(ApplicationID appId)
+    {
+        List<AxisRef> axisRefs = getReferenceAxes(appId)
+        Set<ApplicationID> uniqueAppIds = getReferenceAxesAppIds(axisRefs.toArray(), false)
+        Map<String, ApplicationID> checklist = [:]
+        for (ApplicationID refAppId : uniqueAppIds)
         {
+            if (refAppId.status == ReleaseStatus.SNAPSHOT.name())
```
Comment (jdereg) [inline src/main/groovy/com/cedarsoftware/ncube/NCubeManager.groovy]: ReleaseStatus.SNAPSHOT.name() is an invariant.  Please move outside loop.

File: src/main/groovy/com/cedarsoftware/controller/NCubeController.groovy [hunk 2]
```diff
@@ -767,20 +769,20 @@ class NCubeController implements NCubeConstants, RpmVisualizerConstants
...
-        args[ReferenceAxisLoader.REF_BRANCH] = refAppId.branch
-        args[ReferenceAxisLoader.REF_CUBE_NAME] = refCubeName  // cube name of the holder of the referring (pointing) axis
-        args[ReferenceAxisLoader.REF_AXIS_NAME] = refAxisName    // axis name of the referring axis (the variable that you had missing earlier)
+        args[REF_TENANT] = refAppId.tenant
+        args[REF_APP] = refAppId.app
+        args[REF_VERSION] = refAppId.version
+        args[REF_STATUS] = refAppId.status
+        args[REF_BRANCH] = refAppId.branch
+        args[REF_CUBE_NAME] = refCubeName  // cube name of the holder of the referring (pointing) axis
+        args[REF_AXIS_NAME] = refAxisName    // axis name of the referring axis (the variable that you had missing earlier)
```
Comment (gaigstroliac) [inline src/main/groovy/com/cedarsoftware/controller/NCubeController.groovy]: Just out of curiosity, is there a reason why we couldn't/don't write this like:

```groovy
Map args = [
  (REF_TENANT) : refAppId.tenant,
  (REF_APP)    : refAppId.app,
  (REF_VERSION): refAppId.version, 
...
]
```

File: src/main/groovy/com/cedarsoftware/ncube/NCubeManager.groovy [hunk 3]
```diff
@@ -785,22 +793,65 @@ class NCubeManager implements NCubeMutableClient, NCubeTestServer
...
+            {
+                uniqueAppIds.add(srcApp)
+            }
+            ApplicationID destAppId = new ApplicationID(srcApp.tenant, axisRef.destApp, axisRef.destVersion, axisRef.destStatus, axisRef.destBranch)
             ApplicationID.validateAppId(destAppId)
             assertPermissions(destAppId, axisRef.destCubeName)
+            if (!source)
+            {
+                uniqueAppIds.add(destAppId)
+            }
```
Comment (jsnyder4) [inline src/main/groovy/com/cedarsoftware/ncube/NCubeManager.groovy]: I guess I didn't see this before. Can the if statements on lines 796 and 803 just be combined into an if-else?

File: src/main/groovy/com/cedarsoftware/ncube/NCubeManager.groovy [hunk 4]
```diff
@@ -785,22 +793,65 @@ class NCubeManager implements NCubeMutableClient, NCubeTestServer
...
+                    uniqueAppIds.add(transformAppId)
+                }
             }
         }
+        return uniqueAppIds
+    }
 
-        // Make sure we are not lock blocked on any of the appId's that are being updated.
-        for (ApplicationID appId : uniqueAppIds)
+    private boolean validateReferenceAxesAppIds(ApplicationID appId)
```
Comment (jsnyder4) [inline src/main/groovy/com/cedarsoftware/ncube/NCubeManager.groovy]: Does this method even need a return type? It don't see where we're actually using the return boolean. It looks like it's just a "gate" that throws exceptions if necessary.

**Pull request merged**